cloudify/utils.py

```diff
@@ -1058,3 +1063,53 @@ def uuid4():
         uuid_as_hex[16:20],
         uuid_as_hex[20:]
     )
+
+
+def keep_trying_http(total_timeout_sec=KEEP_TRYING_HTTP_TOTAL_TIMEOUT_SEC,
+                     max_delay_sec=MAX_WAIT_BETWEEN_HTTP_RETRIES_SEC):
+    """
+    Keep (re)trying HTTP requests.
+
+    This is a wrapper function that will keep (re)running whatever function it
+    wraps until it raises a "non-repairable" exception (where "repairable"
+    exceptions are `requests.exceptions.ConnectionError` and
+    `requests.exceptions.Timeout`), or `total_timeout_sec` is exceeded.
+
+    :param total_timeout_sec: The amount of seconds to keep trying, if `None`,
+                              there is no timeout and the wrapped function
+                              will be re-tried indefinitely.
+    :param max_delay_sec: The maximum delay (in seconds) between consecutive
+                          calls to the wrapped function.  The actual delay will
+                          be a pseudo-random number between
+                          `0` and `max_delay_sec`.
+    """
+    logger = setup_logger('http_retrying')
+
+    def ex_to_str(exception):
+        return str(exception) or str(type(exception))
+
+    def inner(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            timeout_at = None if total_timeout_sec is None \
+                else datetime.utcnow() + timedelta(seconds=total_timeout_sec)
+            while True:
+                try:
+                    return func(*args, **kwargs)
+                except (ConnectionError, Timeout) as ex:
+                    if timeout_at and datetime.utcnow() > timeout_at:
+                        logger.info(f'Finished retrying {func}: '
+                                    f'total timeout of {total_timeout_sec} '
+                                    f'seconds exceeded: {ex_to_str(ex)}')
+                        raise
+                    delay = random.randint(0, max_delay_sec)
+                    logger.info(f'Will retry {func} in {delay} seconds: ' +
```

Comment:
This should probably be either warn or debug, not info. I'm not sure which- depends on how often we're expecting this to be and whether we consider it to be something that is at all remarkable.
I'm leaning towards warn.